upd765: sense interrupt status and i82072 changes

pmackinlay opened the pull request:
* changed sense interrupt status drive busy bit handling for seek/recalibrate (only cleared when first result byte is read)
* changed sis to return results in decrementing drive order
* added i82072 motor on/off command
* switched to logmacro.h logging convention

Note: I haven't tested the sense interrupt status change on anything other than the InterPro systems I'm targeting. Hopefully someone else is set up to do so: only the change to the order of sis results is dubious - it's not covered by the documentation I've seen so far, but relied on by the InterPro 2000 diagnostic code.

cracyc commented:
The reverse order from irq sense makes the PC/XT very unhappy.  It checks the order the polling irqs arrive  after a controller reset.

pmackinlay commented:
Thanks for the feedback. I guess you mean that the PC code checks the sense responses and compares the drive number to something? In other words, does it explicitly check that the first response is for drive number 0?

If so, the only option left that I can see is that non drive poll interrupts clear all pending drive poll interrupts. I can't see how a LIFO priority system could work assuming drive 0 is always polled first and drive 3 last (this is explicit in the 82077aa documentation at least), and this was my only other thought.

cracyc commented:
>assuming drive 0 is always polled first and drive 3 last
I've used a logic analyser on a the drive select pins of a zilog 765 clone and can confirm this.  I can try to see what happens to the polling status after a seek or any other behavior if you think that'd help.

pmackinlay commented:
@cracyc, thanks for the offer and info. I've updated this PR with different behaviour that hopefully will not upset the PC/XT; essentially it now looks for any non-drive poll results and returns those first. If there are only drive poll results, they'll be returned in the original drive 0-3 order. This implementation is pretty unsatisfactory to me, but it might by chance mirror what's really going on.

My current theory is that the drive poll interrupt (and the associated results) are cleared at some point when they're "ignored". It's not clear what causes the INT line to be de-asserted in this situation, although the i82077aa datasheet has the following to say: "It should be noted that although four SENSE INTERRUPT STATUS commands are issued, the INT pin is only active until the first SENSE INTERRUPT STATUS command is executed".

Perhaps starting a command other than SIS (recalibrate/seek in my case) has a side-effect of clearing any existing pending drive poll results? This still leaves a fairly big hole given that drive polling can occur during the seek/recalibrate operation, and I have no idea which result SIS would return first in that case.

Not sure if that's something that you could apply your logic analyser to, i.e. trying to figure out what (other than SIS) clears INT after a drive poll interrupt?

cracyc commented:
I can tell you from testing that anything written to the command register, just a single byte, will clear the interrupt line (this is heavily abused by the x68000).   Also, seek results will overwrite poll results for a given drive.  In other words, if the controller is reset then a seek issued, sis will return the poll results as usual except the seeking drive will show seek results in the 0-3 order.

pmackinlay commented:
This leaves me out of ideas. The InterPro does a reset (which triggers poll on all 4 drives because RDY isn't connected), followed by recalibrate of drive 3, followed by sis. There's no sis after the poll interrupt in this case (although there are the expected 4 in other cases), but the single sis after recalibrate completes is expected to give the recalibrate results? Any ideas about how the recalibrate result is prioritized (or clearing the pending drive poll results) would be very welcome.

cracyc commented:
Two possibilities I can think of:
1. I've tested using a 765, maybe the 82072 behaves as you suggest.
2. Polling is disabled with the config command but the reset clears that (lock doesn't seem to exist on the 82072?).

pmackinlay commented:
Can you test this current PR on PC/XT and let me know if it breaks anything? I still think the most probable logic is that the seek/recalibrate command sequence clears the drive poll state at some point. In your earlier comment you said a seek after reset will return poll results as well as the seek result for the specific drive, but I'd be interested to know if that's because the interrupt being serviced is actually due to the poll or seek completion?

Perhaps if the seek were to complete prior to the poll, and therefore be the cause of the interrupt, only one sis result would be present?

I think what's required here is to test the actual chip behavior under the same scenario, that is:

1. Reset
2. Wait for and ignore poll interrupt
3. Execute recalibrate or seek
4. Wait for interrupt and execute sis
5. For extra points, execute another sis

Step 3 would have to be a drive other than 0, and we'd be interested to see whether step 5 returns a valid result or error.

I guess this is something that might be able to be tested on a VM, but best would be on real hardware (which I don't have unfortunately).

cracyc commented:
The XT seems okay with this but I tried your test on my 765 and on a pc87312 superio which claims full 82077 compatibility and on both the first sis result after a recalibrate of drive 3 was ready status for drive 0 and the second ready for drive 1 etc...  Can this be tested in the emulation by just running the IPL or does it need to have a floppy to boot?

pmackinlay commented:
I've carefully reviewed the boot sequence on my system again, and the only thing I can see that might have an impact is that it actually executes a specify command in between steps 2 and 3. Can you possibly retry the same test including a specify and see if this changes anything? I retain a fading hope that this might be clearing the pending drive poll results...

So, more precisely, here's what the InterPro 2000 boot code does:

1. Reset by setting DSR SWR (bit 7)
2. Wait for interrupt (but don't send any SIS commands)
3. Execute specify 0x83 0x80
4. Recalibrate drive 3
5. Wait for interrupt
6. Execute SIS, expect 0x23 0x00

This testing only requires the boot code (IPL) in the InterPro - unfortunately, it needs changes to the InterPro driver that I haven't yet pushed because there are so many open problems I'm working on.

cracyc commented:
Nope, same.  One thing I did notice though is the busy bit isn't cleared until the results for the seeking drive are read.  I don't know if that makes a difference.  Otherwise, unless polling is disabled which is only other way i can see that that would happen, IMO those changes should be made specific for the controller type so you aren't blocked from progress and it can be revisited later.

pmackinlay commented:
@cracyc, please let me know if this version works with PC/XT. I believe it will and that it's relatively safe because the logic changes are now largely limited to the i82072.

cracyc commented:
Yes, that's okay.  I'll wait a bit for any other comments but it's fine with me to be committed.

cuavas commented on src/devices/machine/upd765.cpp:
```diff
@@ -267,14 +270,18 @@ bool upd765_family_device::get_ready(int fid)
+		return;
+
+	// pass drive select to connected drives
+	for (auto &fi : flopi)
```
Can you please not use `auto` like this?
* You definitely know what the type of object you want a reference to is
* The name is not excessively long or deeply qualified
* The type is not dependent of template arguments
* It is not an iterator or proxy to the actual object

Using `auto` here just makes the code less approachable.

cuavas commented on src/devices/machine/upd765.cpp:
```diff
@@ -376,6 +383,18 @@ READ8_MEMBER(upd765_family_device::msr_r)
+	// log msr when changed
+	if (VERBOSE)
+	{
+		static u8 last_msr = 0;
```
Oh no, this is a deal-breaker.  No non-const function statics.  It breaks save state support and systems with multiple instances of the same device.

cuavas commented on src/devices/machine/upd765.cpp:
```diff
@@ -376,6 +383,18 @@ READ8_MEMBER(upd765_family_device::msr_r)
 		}
 
+	// log msr when changed
+	if (VERBOSE)
```
This file appears to be mostly K&R style - please follow that with additions.

cuavas commented on src/devices/machine/upd765.cpp:
```diff
@@ -414,9 +433,16 @@ READ8_MEMBER(upd765_family_device::fifo_r)
 		memmove(result, result+1, result_pos);
 		if(!result_pos)
 			main_phase = PHASE_CMD;
+		else if (result_pos == 1)
```
More Allman creeping into a K&R file...

cuavas commented on src/devices/machine/upd765.cpp:
```diff
@@ -414,9 +433,16 @@ READ8_MEMBER(upd765_family_device::fifo_r)
+		else if (result_pos == 1)
+		{
+			// clear drive busy bit after the first sense interrupt status result byte is read
+			for (auto &fi : flopi)
```
Another place `auto` type isn't helpful

cuavas commented on src/devices/machine/upd765.cpp:
```diff
@@ -2473,6 +2510,186 @@ i82072_device::i82072_device(const machine_config &mconfig, const char *tag, dev
+		 */
+
+		// clear pending interrupts and fall through
+		for (auto &fi : flopi)
```
More inappropriate `auto`

cuavas commented on src/devices/machine/upd765.h:
```diff
@@ -332,11 +332,12 @@ class upd765_family_device : public pc_fdc_interface {
 	uint8_t dor, dsr, msr, fifo[16], command[16], result[16];
 	uint8_t st1, st2, st3;
-	uint8_t fifocfg, dor_reset;
+	uint8_t fifocfg, motorcfg, dor_reset;
```
Shouldn't the new members be registered for save state?

pmackinlay commented:
@cuavas  thanks for the feedback. Hopefully addressed all points, however note that this device doesn't have save state for anything other than the new motorcfg member. I am not familiar enough with the rest of the device to be sure I wouldn't get that wrong.

cuavas commented:
You've still added Allman style code to a K&R file.  Inconsistent code is worse than (subjectively) ugly code.

The file is written like:
```
if(x) {
```
(i.e. no space between keyword and opening parenthesis, opening brace at end of line)

But you've added code like this:
```
if (x)
{
```
(i.e. space between opening parenthesis and/or opening brace on its own line)

Could you please make it all consistent with the code already in the file?

pmackinlay commented:
@cuavas, all done now. I also corrected a few non-K&R instances in code I'd otherwise not touched, so hopefully it's more consistent overall.

Tafoid commented:
https://github.com/mamedev/mame/commit/5a3081ed8ebf74e10942e0382b106ad6a2d459fb
http://mametesters.org/view.php?id=6689
cpc6128 attempted CAT of mounted disk causes MAME to hardlock/hang.  Other drivers may also be affected.
